iquip/apps/logger.py

```diff
@@ -0,0 +1,188 @@
+"""App module for log viewer in apps."""
+
+import logging
+from typing import Any, Optional, Tuple, Callable
+
+from PyQt5.QtCore import QObject, pyqtSlot, pyqtSignal, QDateTime
+from PyQt5.QtWidgets import (
+    QWidget, QVBoxLayout, QPushButton, QTextEdit, QLabel, QDialogButtonBox, QComboBox
+)
+
+import qiwis
+
+logger = logging.getLogger(__name__)
+
+
+class _Signaller(QObject):
+    """Signal only for LoggingHandler.
+
+    Signals:
+        signal(log): A formatted log message is emitted. 
+    """
+
+    signal = pyqtSignal(str)
+
+
+class LoggingHandler(logging.Handler):
+    """Handler for logger.
+
+    Sends a log message to the connected function through a signal.   
+    """
+
+    def __init__(self, slotfunc: Callable[[str], Any]):
+        """Extended.
+
+        Connects the slotfunc to the signal.
+
+        Args:
+            slotfunc: A slot function which is called when a log record is emitted.
+        """
+        super().__init__()
+        self.signaller = _Signaller()
+        self.signaller.signal.connect(slotfunc)
+
+    def emit(self, record: logging.LogRecord):
+        """Overridden.
+        
+        Emits input signal to the connected function.
+        """
+        logMsg = self.format(record)
+        self.signaller.signal.emit(logMsg)
+
+
+class LoggerFrame(QWidget):
+    """Frame for logging.
+
+    Attributes:
+        logEdit: A textEdit which shows all logs.
+        clearButton: A button for clearing all logs.
+        levelBox: A comboBox for setting the logger's level.
+    """
+
+    def __init__(self, parent: Optional[QObject] = None):
+        """Extended."""
+        super().__init__(parent=parent)
+        # widgets
+        self.logEdit = QTextEdit(self)
+        self.logEdit.setReadOnly(True)
+        self.clearButton = QPushButton("Clear", self)
+        self.levelBox = QComboBox(self)
+        # layout
+        layout = QVBoxLayout(self)
+        layout.addWidget(self.logEdit)
+        layout.addWidget(self.clearButton)
+        layout.addWidget(self.levelBox)
+
+
+class ConfirmClearingFrame(QWidget):
+    """A confirmation frame for log clearing in the LoggerFrame.
+    
+    Attributes:
+        label: Displays a confirmation message to clear logs in the LoggerFrame.
+        buttonBox: Contains OK and Cancel button to check whether to clear logs.
+    """
+
+    confirmed = pyqtSignal()
```

Comment:
Please add the docstring about this signal.

Reply:
You should add "Signals:" section in the docstring.